### PR TITLE
Added support for multi dimensional array.

### LINKED ISSUE #1600: Support multi dimensional arrays
Currently we only support one dimensional arrays in KSQL. We need to able to support multidimensional arrays too.

### PR DESCRIPTION
**Pull request by hjafarpour**
### Description 
This PR adds multi dimensional array support.

Fixes #1600 

### Testing done 
Unit and integration tests were added.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-engine/src/test/resources/query-validation-tests/project-filter.json [hunk 1]
```diff
@@ -94,6 +94,27 @@
           "timestamp": 0
         }
       ]
+    },
+    {
+      "name": "Json Multi Dimensional Array",
+      "statements": [
+        "CREATE STREAM array_array (ID BIGINT, ARRAY_COL ARRAY<ARRAY<VARCHAR>>) WITH (kafka_topic='test_topic', value_format='JSON');",
+        "CREATE STREAM S3 as SELECT ID, ARRAY_COL[0][1] AS array_item FROM array_array;"
```
Comment (apurvam) [inline ksql-engine/src/test/resources/query-validation-tests/project-filter.json]: What's the error if the `SELECT` is on `ARRAY_COL[0][3]`? Do we log it? Do we just log an `ArrayOutOfBoundsException`? IF the user is doing this on a CLI, what do they see? 

This is lacking negative test cases or any discussion of the UX for the failing path.

Comment (hjafarpour) [reply ksql-engine/src/test/resources/query-validation-tests/project-filter.json]: The out of bound error is handled the same way we handle other errors while computing the expressions. We log the error and return `null`.

**Comment by blueedgenick** [comment]
I wonder if this is the best way to solve the requirement ? A couple of thoughts: (1) looks like this only handles 2-dimensional arrays, as opposed to genuinely n-dimensional ones ? (at least that's what the test is testing). I know that was the use-case raised in the issue filed on GH but seems like we should have a general solution; (2) a more SQL-y way is to use a function here rather than a `col1[n][n][n][n]...[n]` syntax. If we added a simple `element_at(array-col, index)` then the user could nest calls to it as deep as she wanted without problem and it would work in an intuitively similar way to any other dereferencing function ?

File: ksql-engine/src/test/resources/query-validation-tests/project-filter.json (hunk 1 above)
Comment (apurvam) [reply ksql-engine/src/test/resources/query-validation-tests/project-filter.json]: What will the user see on the CLI? Is it a useful message? Is there a test for the negative case that exemplifies this message?

**Comment by hjafarpour** [comment]
@blueedgenick About the first point you raise, the implementation is generic and I just added a test for 3 dimensional array. You can have any number of dimensions based on the current implementation.
As for using function instead of dimensions I think both options are valid but the current syntax is more intuitive IMO and it also is similar to the way other systems such as postgres(https://www.postgresql.org/docs/9.1/static/arrays.html) deal with multidimensional arrays.
By the way, I think we should definitely add array functions and operators similar to the ones in presto (https://prestodb.io/docs/current/functions/array.html).

**Comment by blueedgenick** [comment]
cool, thanks Hojjat! I don't see any harm in being able to do it either way (loadsa-subscripts or function-based), provided they both work for n-dimensions.
Re. array functions, a la Presto etc. - funny you should mention that, i have a PR brewing.... ;)

File: ksql-engine/src/test/resources/query-validation-tests/project-filter.json (hunk 1 above)
Comment (hjafarpour) [reply ksql-engine/src/test/resources/query-validation-tests/project-filter.json]: No, user won't see any messages since this is per row. The error message along with the exception details will be in the log.

**Comment by hjafarpour** [comment]
@apurvam wanna take another look at this?

**Comment by apurvam** [comment]
@hjafarpour , I think this is still missing the test for the negative case which I asked for originally. When we get an `ArrayIndexOutOfBounds` for a given dereference, we should be setting that column value to `null`, but we don't test for this.
